recipes/pyspark-stubs/meta.yaml

```diff
@@ -0,0 +1,43 @@
+{% set name = "pyspark-stubs" %}
+{% set version = "2.4.0.post5" %}
+
+package:
+  name: "{{ name|lower }}"
+  version: "{{ version }}"
+
+source:
+  url: "https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz"
+  sha256: "2249e20f3d120551d6ba3bf8bca67c1cb9a2d2e39bd9ebcd3cb00ae906428f21"
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv "
+  skip: true  # [py<36]
+
+requirements:
+  host:
+    - pip
+    - pyspark >=2.4.0,<3.0.0
+    - python
+  run:
+    - pyspark >=2.4.0,<3.0.0
+    - python
+
+test:
+  source_files:
+    - third_party
+  commands:
+    - pip install git+https://github.com/python/mypy.git@master && mypy third_party/3/pyspark
```

Comment:
I think it is better if you use ``mypy`` from the ``conda-forge`` channel
that should be
```
  requires:
    - mypy
```

Reply:
I thought about it, but it might not work well in practice.  For different reasons I test annotations against mypy development branch. In practice it means that:

- Certain constructs might not work against latest release.
- Some mypy bug fixes, require significant changes in annotations, which are not necessarily backwards compatible.

Reply:
Using latest master as a test dependency will lead to irreproducible results.
Can you use a certain dev release or at least an specific commit?

Reply:
@ocefpaf That's a valid concern, though I am not sure if it is that much different from having a dependence without version bound or only with a lower bound. In such cases one can end up with unreproducible test results as well. 

In general the source of the problem is that mypy, as de facto gold standard for Python type checker, is used as an acceptance test, it is not a dependency for the package, nor it is assumed that it is used by the final user (or even that the final user utilizes compatible tool).


As of possible resolution:

- AFAIK mypy doesn't tag or release pre-releases, but specific commit is definitely an option.

- An alternative is to bind to some specific version, and either use preceding release, or the following one if the preceding doesn't work. However the latter option would delay forge release, which is not optimal.

- Finally it should be fine to skip this test, however there is very little than can be tested in annotation-only package otherwise. Maybe file existence, as even imports are not meaningful. 

I haven't seen any stub packages in the feedstocks, so I guess it is fairly generic problem - how such packages, if contributed in the future, should be tested?

Reply:
> @ocefpaf That's a valid concern, though I am not sure if it is that much different from having a dependence without version bound or only with a lower bound. In such cases one can end up with unreproducible test results as well.

While you are correct it is easier to inspect the logs and see the version used in the build than the commit pulled from master.

> In general the source of the problem is that mypy, as de facto gold standard for Python type checker, is used as an acceptance test, it is not a dependency for the package, nor it is assumed that it is used by the final user (or even that the final user utilizes compatible tool).

Got it. That raises another question. Do we need that test on the package then? Maybe it is a good test for upstream code testing, right?

> As of possible resolution:
> 
> * AFAIK mypy doesn't tag or release pre-releases, but specific commit is definitely an option.
> * An alternative is to bind to some specific version, and either use preceding release, or the following one if the preceding doesn't work. However the latter option would delay forge release, which is not optimal.
> * Finally it should be fine to skip this test, however there is very little than can be tested in annotation-only package otherwise. Maybe file existence, as even imports are not meaningful.
> 
> I haven't seen any stub packages in the feedstocks, so I guess it is fairly generic problem - how such packages, if contributed in the future, should be tested?

I guess that testing for an existing file is OK. I'm not against merging this as-is as well. Based on your explanation it is not a big deal.

Reply:
> While you are correct it is easier to inspect the logs and see the version used in the build than the commit pulled from master.
> 
Fair enough.

> Got it. That raises another question. Do we need that test on the package then? Maybe it is a good test for upstream code testing, right?
> 
Because I control PyPI release process, where passing mypy and data driven tests are pre-requisites for release, and this recipe just piggy-backs on top of PyPI, I'd say that's the case. 

And even if the tests fail here, it is more indicator of losing sync with the ecosystem (mypy, typshed) than a problem with the recipe.

> I guess that testing for an existing file is OK. I'm not against merging this as-is as well. Based on your explanation it is not a big deal.

So maybe checking that installation has been patched with `py.typed`, i.e.

```yaml
test:
  commands:
    - test -f $SP_DIR/pyspark/py.typed  # [unix]
    - if not exist %SP_DIR%\\pyspark\\py.typed exit 1  # [win]
```
is sufficient?

Reply:
> So maybe checking that installation has been patched with `py.typed`, i.e.

IMO yes. That would suffice. In the future, if that approach fails, we can re-assess our strategy here.

Reply:
Then it is decided. I've [modified](https://github.com/conda-forge/staged-recipes/pull/9480/commits/80fd7aa9d39d6318860c02313b879983e06665e9) the recipe/

Thank you for all your comments @ocefpaf!